sparse/dok.py

```diff
@@ -1,31 +1,21 @@
-import six
+from builtins import dict, int
```

Comment:
What is `dict` here in Python 2?  It's not a normal builtin dict object?

Ah, ok, I'm now reading about the `future` package.  I have to admit that I am a bit hesitant to add it as a new dependency just because I haven't heard of it before.  Happy to deal with this in the future though.

Reply:
`future` is a dependency for many packages. It's more common than `six`, I believe. I may be wrong.

It's `pip` installable.

Reply:
> future is a dependency for many packages. It's more common than six, I believe. I may be wrong.

This would make me much more comfortable.  I would be somewhat surprised to learn this though.

I hesitate to have a `dict` type anywhere in the code that is not the standard builtin dict type, both for performance and (more importantly) consistency reasons with the ecosystem.

Reply:
I just took a peek through the code for `future`.

```
    def items(self):
        """
        On Python 2.7+:
            D.items() -> a set-like object providing a view on D's items
        On Python 2.6:
            D.items() -> an iterator over D's items
        """
        if ver == (2, 7):
            return self.viewitems()
        elif ver == (2, 6):
            return self.iteritems()
        elif ver >= (3, 0):
            return self.items()
```

Reply:
What if some other infrastrucural library (like cloudpickle, Cython, PyPy, Numba, ...) has an optimiziation that works well on objects where `type(obj) is dict` (this is almost certainly the case in all of those projects).  Then presumably that optimization won't apply to the DOK class.

This library is aimed to be infrastructural for the ecosystem.  I think that it is important that we be as non-fancy as possible.  As another point, if you ever want to see this upstreamed into `scipy` then you'll need to convince them about this point as well and they are, in general, far more conservative than I am.

Reply:
I had this blogpost sitting on my hard drive as a draft; I just pushed it to my website.  

http://matthewrocklin.com/blog/work/2018/01/27/write-dumb-code

It has my thoughts on the topic of avoiding technology in programming whenever possible.  I may be a bit dogmatic on this topic, I apologize if this causes friction.

Reply:
`builtins` basically aims to map Python 3 behavior everywhere. It also has a `past` module for mapping Python 2 behavior anywhere.

Reply:
Yeah, it's a nice trick, don't get me wrong.  I've just seen many many things go wrong once code like this interacts with 1000 other libraries (which again, I think is our intention).  I would definitely use a library like this in application code, but for core ecosystem code I tend to be more conservative.  I think that many other maintainers of core ecosystem code also hold this belief.

Reply:
Hmm. I'd be okay going back to `six`, but the memory constrains could get bad. What do you think?

Reply:
> I apologize if this causes friction.

Not at all. :-) We're just discussing.

Reply:
I would wait for someone to arrive with a memory issue before optimizing for memory constraints in this way.  I don't expect this to be a significant issue in common use.  Also, I don't mind Python 2 being less efficient.

Reply:
I removed the `six` dependency entirely and added our own basic Python 2/3 compat.